source/common/config/config_provider_impl.h

```diff
@@ -0,0 +1,370 @@
+//      be called to fetch either an existing ConfigSubscriptionInstanceBase if the config source
+//      configuration matches, or a newly instantiated subscription otherwise.
+//
+// For static providers:
+//   1) Create a class derived from StaticConfigProviderImplBase and implement the required
+//   interface.
+//
+// For dynamic (xDS) providers:
+//   1) Create a class derived from DynamicConfigProviderImplBase and implement the required
+//   interface. 2) Create a class derived from ConfigSubscriptionInstanceBase; this is the entity
+//   responsible for owning and managing the Envoy::Config::Subscription<ConfigProto> that provides
+//   the underlying config subscription.
+//     - When subscription callbacks (onConfigUpdate, onConfigUpdateFailed) are issued by the
+//     underlying subscription, the corresponding ConfigSubscriptionInstanceBase functions must be
+//     called as well.
+//     - On a successful config update, checkAndApplyConfig() should be called to instantiate the
+//     new config implementation and propagate it to the shared config providers and all
+//     worker threads.
+
+class ConfigProviderManagerImplBase;
+
+/**
+ * ConfigProvider implementation for statically specified configuration.
+ *
+ * This class can not be instantiated directly; instead, it provides the foundation for
+ * static config provider implementations which derive from it.
+ */
+class StaticConfigProviderImplBase : public ConfigProvider {
+public:
+  StaticConfigProviderImplBase() = delete;
+
+  virtual ~StaticConfigProviderImplBase();
+
+  // Envoy::Config::ConfigProvider
+  SystemTime lastUpdated() const override { return last_updated_; }
+
+protected:
+  StaticConfigProviderImplBase(Server::Configuration::FactoryContext& factory_context,
+                               ConfigProviderManagerImplBase& config_provider_manager);
+
+private:
+  SystemTime last_updated_;
+  ConfigProviderManagerImplBase& config_provider_manager_;
+};
+
+class DynamicConfigProviderImplBase;
+
+/**
+ * Provides generic functionality required by all dynamic ConfigProvider subscriptions, including
+ * shared lifetime management via shared_ptr.
+ *
+ * To do so, this class keeps track of a set of dynamic config providers associated with an
+ * underlying subscription; providers are bound/unbound as needed as they are created and destroyed.
+ * This is done to avoid exposing shared ownership semantics to the external API (see
+ * ConfigProvider).
+ *
+ * This class can not be instantiated directly; instead, it provides the foundation for
+ * dynamic config subscription implementations which derive from it.
+ */
+class ConfigSubscriptionInstanceBase : public Init::Target,
+                                       protected Logger::Loggable<Logger::Id::config> {
+public:
+  struct LastConfigInfo {
+    uint64_t last_config_hash_;
+    std::string last_config_version_;
+  };
+
+  ConfigSubscriptionInstanceBase() = delete;
```

Comment:
Right, it's just my preference to explicitly state the intent of the class by deleting the default constructor.

Now that the class has been renamed, the intent is fairly clear and like you said, compilation will fail when improperly instantiated due to the pure virtual functions, so I'm ok removing this since this doesn't appear to be a common idiom across the codebase.

Reply:
Constructor deletes have been removed.